File: classes/barbarian.py
from typing import List, Optional

import sim.core_feats
from util.util import get_magic_weapon, apply_asi_feats
from feats.epic_boons import IrresistibleOffense
from feats.origin import SavageAttacker
from feats import (
    ASI,
    GreatWeaponMaster,
    PolearmMaster,
    AttackAction,
    WeaponMasteries,
)
from weapons import Glaive, Greatsword, GlaiveButt

import sim.feat
import sim.character
import sim.weapons


class BarbarianLevel(sim.core_feats.ClassLevels):
    def __init__(self, level: int):
        super().__init__(name="Barbarian", level=level)


class Frenzy(sim.feat.Feat):
    def __init__(self, num_dice: int):
        self.used = False
        self.num_dice = num_dice

    def begin_turn(self, target):
        self.used = False

    def attack_result(self, args):
        if args.hits() and not self.used:
            self.used = True
            args.add_damage(source="Berserker", dice=self.num_dice * [6])


class BrutalStrike(sim.feat.Feat):
    def __init__(self, num_dice: int):
        self.num_dice = num_dice

    def begin_turn(self, target):
        self.used = False

    def attack_roll(self, args):
        if not self.used and args.adv:
            args.adv = False
            args.attack.add_tag("brutal_strike")
            self.used = True

    def attack_result(self, args):
        if args.hits() and args.attack.has_tag("brutal_strike"):
            args.add_damage(source="BrutalStrike", dice=self.num_dice * [10])


class Retaliation(sim.feat.Feat):
    def __init__(self, weapon):
        self.weapon = weapon

    def enemy_turn(self, target):
        self.character.weapon_attack(target, weapon=self.weapon)


class PrimalChampion(sim.feat.Feat):
    def apply(self, character):
        super().apply(character)
        character.increase_stat_max("str", 4)
        character.increase_stat_max("con", 4)
        character.increase_stat("str", 4)
        character.increase_stat("con", 4)


class Rage(sim.feat.Feat):
    def __init__(self, dmg):
        self.raging = False
        self.dmg = dmg

    def short_rest(self):
        self.raging = False

    def before_action(self, target):
        if not self.raging and self.character.use_bonus("rage"):
            self.raging = True

    def attack_result(self, args):
        if args.hits() and self.raging:
            args.add_damage(source="Rage", damage=self.dmg)


class RecklessAttack(sim.feat.Feat):
    def begin_turn(self, target):
        self.enabled = True

    def attack_roll(self, args):
        if self.enabled:
            args.adv = True

    def end_turn(self, target):
        self.enabled = False


def rage_damage(level: int):
    if level >= 16:
        return 4
    elif level >= 9:
        return 3
    return 2


def barbarian_feats(
    level: int, asis: Optional[List["sim.feat.Feat"]] = None
) -> List["sim.feat.Feat"]:
    rage_dmg = rage_damage(level)
    feats: List["sim.feat.Feat"] = []
    if level >= 1:
        feats.append(BarbarianLevel(level))
        feats.append(WeaponMasteries(["Graze", "Topple"]))
        feats.append(Rage(dmg=rage_dmg))
    # Level 1 (Unarmored Defense) is irrelevant
    if level >= 2:
        feats.append(RecklessAttack())
    # Level 3 (Primal Knowledge) is irrelevant
    # Level 5 (Fast Movement) is irrelevant
    # Level 7 is irrelevant
    if level >= 9:
        feats.append(BrutalStrike(num_dice=2 if level >= 17 else 1))
    # Level 11 is irrelevant
    # Level 15 is irrelevant
    # Level 18 is irrelevant (until Strength checks or saves matter)
    if level >= 20:
        feats.append(PrimalChampion())
    apply_asi_feats(level, feats, asis)
    return feats


def berserker_feats(level: int, weapon: "sim.weapons.Weapon") -> List["sim.feat.Feat"]:
    rage_dmg = rage_damage(level)
    feats: List["sim.feat.Feat"] = []
    if level >= 3:
        feats.append(Frenzy(num_dice=rage_dmg))
    # Level 6 is irrelevant
    if level >= 10:
        feats.append(Retaliation(weapon))
    # Level 14 is irrelevant
    return feats


class PolearmBarbarian(sim.character.Character):
    def __init__(self, level: int):
        magic_weapon = get_magic_weapon(level)
        feats: List["sim.feat.Feat"] = []
        weapon = Glaive(magic_bonus=magic_weapon)
        feats.append(AttackAction(attacks=[weapon, weapon] if level >= 5 else [weapon]))
        feats.extend(
            barbarian_feats(
                level,
                asis=[
                    GreatWeaponMaster(weapon),
                    PolearmMaster(GlaiveButt(magic_bonus=magic_weapon)),
                    ASI(["dex", "str"]),
                    ASI(["dex"]),
                    IrresistibleOffense("str"),
                ],
            )
        )
        feats.extend(berserker_feats(level, weapon))
        feats.append(SavageAttacker())
        super().init(
            level=level,
            stats=[17, 10, 10, 10, 10, 10],
            base_feats=feats,
        )


class BerserkerBarbarian(sim.character.Character):
    def __init__(self, level: int, **kwargs):
        magic_weapon = get_magic_weapon(level)
        feats: List["sim.feat.Feat"] = []
        weapon = Greatsword(magic_bonus=magic_weapon)
        feats.append(AttackAction(attacks=[weapon, weapon] if level >= 5 else [weapon]))
        feats.extend(
            barbarian_feats(
                level,
                asis=[
                    GreatWeaponMaster(weapon),
                    ASI(["str"]),
                    ASI(["dex"]),
                    ASI(["dex"]),
                    IrresistibleOffense("str"),
                ],
            )
        )
        feats.extend(berserker_feats(level, weapon))
        feats.append(SavageAttacker())
        super().init(
            level=level,
            stats=[17, 10, 10, 10, 10, 10],
            base_feats=feats,
        )

File: src/classes/Barbarian.ts
import { Environment } from "../sim/Environment"
import { Operation } from "../sim/actions/Operation"
import { Character } from "../sim/Character"
import { Feature } from "../sim/Feature"
import { AttackRollEvent } from "../sim/events/AttackRollEvent"
import { WeaponMastery } from "../sim/types"
import { WeaponMasteries } from "../feats/shared/WeaponMasteries"
import { AddClassLevel } from "../sim/coreFeats/ClassLevel"
import { defaultMagicBonus, unreachable } from "../util/helpers"
import { ExtraAttack } from "../feats/shared/ExtraAttack"
import { Resource } from "../sim/resources/Resource"
import { SavageAttacker } from "../feats/origin/SavageAttacker"
import { GreatWeaponMaster } from "../feats/general/GreatWeaponMaster"
import { AbilityScoreImprovement } from "../feats/general/AbilityScoreImprovement"
import { IrresistibleOffense } from "../feats/epic/IrresistibleOffense"
import { SetAttribute } from "../feats/shared/SetAttribute"
import { IncreaseResource } from "../feats/shared/IncreaseResource"
import { AttackResultEvent } from "../sim/events/AttackResultEvent"
import { Effect, EffectDuration } from "../sim/Effect"
import { DefaultAttackActionOperation } from "../operations/DefaultAttackActionOperation"
import { Greatsword } from "../weapons/martial/melee/Greatsword"
import { FeatureGroup } from "../sim/helpers/FeatureGroup"

const RageResource = "rage"
const RageEffectName = "raging"
const RecklessTag = "reckless"

class RageEffect extends Effect {
    name = RageEffectName
    duration: EffectDuration = "until_short_rest"

    apply(character: Character): void {
        character.events.on("attack_result", this.attackResult)
    }

    end(character: Character): void {
        character.events.removeListener("attack_result", this.attackResult)
    }

    attackResult = (event: AttackResultEvent): void => {
        const weapon = event.attack?.attack.weapon()
        if (
            weapon &&
            event.hit &&
            event.attack.attack.stat(this.character) == "Str"
        ) {
            event.addDamage({
                source: "Rage",
                flatDmg: this.character.getAttribute(RageBonusDamageAttribute),
                type: weapon.damageType,
            })
        }
    }
}

class RageOperation implements Operation {
    repeatable = false

    eligible(_environment: Environment, character: Character): boolean {
        // While it is techically possible to rage while raging, at this time
        // there is no benefit to doing so
        return (
            character.hasResource(RageResource) &&
            character.bonus.has() &&
            !character.hasEffect(RageEffectName)
        )
    }

    do(_environment: Environment, character: Character): void {
        character.useResource(RageResource)
        character.bonus.use()
        character.addEffect(new RageEffect())
    }
}

const RageBonusDamageAttribute = "rageBonusDamage"

class Rage extends Feature {
    apply(character: Character) {
        this.addResource()
        character.customTurn.addOperation("before_action", new RageOperation())
    }

    addResource() {
        this.character.resources.set(
            RageResource,
            new Resource({
                name: RageResource,
                character: this.character,
                initialMax: 2,
                incrementOnShortRest: true,
                resetOnLongRest: true,
            })
        )
    }
}

class RecklessAttack extends Feature {
    attackRoll(event: AttackRollEvent) {
        if (event.attack.attack.stat(this.character) == "Str") {
            event.adv = true
            event.attack.addTag(RecklessTag)
        }
    }
}

class PrimalChampion extends Feature {
    apply(character: Character) {
        character.increaseStatAndMax("Str", 4)
        character.increaseStatAndMax("Con", 4)
    }
}

class Frenzy extends Feature {
    used = false

    beginTurn() {
        this.used = false
    }

    attackResult(event: AttackResultEvent) {
        const weapon = event.attack?.attack.weapon()
        if (
            weapon &&
            event.hit &&
            !this.used &&
            this.character.hasEffect(RageEffectName) &&
            event.attack?.hasTag(RecklessTag)
        ) {
            const dice = Array(
                this.character.getAttribute(RageBonusDamageAttribute)
            ).fill(6)
            event.addDamage({
                source: "Frenzy",
                dice,
                type: weapon.damageType,
            })
            this.used = true
        }
    }
}

class DivineFury extends Feature {
    used = false

    beginTurn() {
        this.used = false
    }

    attackResult(event: AttackResultEvent) {
        if (
            event.hit &&
            !this.used &&
            this.character.hasEffect(RageEffectName) &&
            event.attack?.attack.weapon()
        ) {
            event.addDamage({
                source: "DivineFury",
                dice: [6],
                flatDmg: Math.floor(
                    this.character.getClassLevel("Barbarian") / 2
                ),
                // This could also be necrotic instead
                type: "radiant",
            })
            this.used = true
        }
    }
}

type BarbarianSubclass = "Berserker" | "Zealot"

export class Barbarian {
    static operations = {
        RageOperation,
    }

    static features(args: {
        level: number
        asis: Feature[]
        masteries: WeaponMastery[]
        subclass: BarbarianSubclass
    }): Feature[] {
        const { level, asis, masteries, subclass } = args
        const features: Feature[] = []
        if (level >= 1) {
            features.push(Barbarian.level1(level, masteries))
        }
        if (level >= 2) {
            features.push(Barbarian.level2())
        }
        if (level >= 3) {
            features.push(Barbarian.level3(subclass))
        }
        if (level >= 4) {
            features.push(Barbarian.level4(asis[0]))
        }
        if (level >= 5) {
            features.push(Barbarian.level5())
        }
        if (level >= 6) {
            features.push(Barbarian.level6(subclass))
        }
        if (level >= 8) {
            features.push(Barbarian.level8(asis[1]))
        }
        if (level >= 9) {
            features.push(Barbarian.level9())
        }
        if (level >= 12) {
            features.push(Barbarian.level12(asis[2]))
        }
        if (level >= 16) {
            features.push(Barbarian.level16(asis[3]))
        }
        if (level >= 17) {
            features.push(Barbarian.level17())
        }
        if (level >= 19) {
            features.push(Barbarian.level19(asis[4]))
        }
        if (level >= 20) {
            features.push(Barbarian.level20())
        }
        return features
    }

    static level1(level: number, masteries: WeaponMastery[]): Feature {
        return new FeatureGroup([
            new AddClassLevel("Barbarian", level),
            new WeaponMasteries(masteries),
            new SetAttribute(RageBonusDamageAttribute, 2),
            new Rage(),
        ])
    }

    static level2(): Feature {
        return new FeatureGroup([new RecklessAttack()])
    }

    static level3(subclass: BarbarianSubclass): Feature {
        if (subclass === "Berserker") {
            return new FeatureGroup([
                new IncreaseResource(RageResource),
                new Frenzy(),
            ])
        } else if (subclass === "Zealot") {
            return new FeatureGroup([
                new IncreaseResource(RageResource),
                new DivineFury(),
            ])
        } else {
            unreachable(subclass)
        }
    }

    static level4(asi: Feature): Feature {
        return new FeatureGroup([asi])
    }

    static level5(): Feature {
        return new FeatureGroup([new ExtraAttack(2)])
    }

    static level6(subclass: BarbarianSubclass): Feature {
        if (subclass === "Berserker") {
            // Level 6 (Mindless Rage) is ignored
            return new FeatureGroup([new IncreaseResource(RageResource)])
        } else if (subclass === "Zealot") {
            return new FeatureGroup([new IncreaseResource(RageResource)])
        } else {
            unreachable(subclass)
        }
    }

    static level8(asi: Feature): Feature {
        return new FeatureGroup([asi])
    }

    static level9(): Feature {
        return new FeatureGroup([new SetAttribute(RageBonusDamageAttribute, 3)])
    }

    static level12(asi: Feature): Feature {
        return new FeatureGroup([new IncreaseResource(RageResource), asi])
    }

    static level16(asi: Feature): Feature {
        return new FeatureGroup([
            new SetAttribute(RageBonusDamageAttribute, 4),
            asi,
        ])
    }

    static level17(): Feature {
        return new FeatureGroup([new IncreaseResource(RageResource)])
    }

    static level19(asi: Feature): Feature {
        return new FeatureGroup([asi])
    }

    static level20(): Feature {
        return new FeatureGroup([new PrimalChampion()])
    }

    static createBerserkerBarbarian(level: number): Character {
        const character = new Character({
            stats: { Str: 17, Dex: 10, Con: 16, Int: 10, Wis: 10, Cha: 10 },
        })
        const weapon = new Greatsword({ magicBonus: defaultMagicBonus(level) })
        const features = [
            new SavageAttacker(),
            ...Barbarian.features({
                level,
                subclass: "Berserker",
                masteries: ["Topple", "Graze"],
                asis: [
                    new GreatWeaponMaster(weapon),
                    new AbilityScoreImprovement("Str"),
                    new AbilityScoreImprovement("Con"),
                    new AbilityScoreImprovement("Con"),
                    new IrresistibleOffense("Str"),
                ],
            }),
        ]
        features.forEach((feat) => character.addFeature(feat))
        character.customTurn.addOperation(
            "action",
            new DefaultAttackActionOperation(weapon)
        )
        return character
    }

    static createZealotBarbarian(level: number): Character {
        const character = new Character({
            stats: { Str: 17, Dex: 10, Con: 16, Int: 10, Wis: 10, Cha: 10 },
        })
        const weapon = new Greatsword({ magicBonus: defaultMagicBonus(level) })
        const features = [
            new SavageAttacker(),
            ...Barbarian.features({
                level,
                subclass: "Zealot",
                masteries: ["Topple", "Graze"],
                asis: [
                    new GreatWeaponMaster(weapon),
                    new AbilityScoreImprovement("Str"),
                    new AbilityScoreImprovement("Con"),
                    new AbilityScoreImprovement("Con"),
                    new IrresistibleOffense("Str"),
                ],
            }),
        ]
        features.forEach((feat) => character.addFeature(feat))
        character.customTurn.addOperation(
            "action",
            new DefaultAttackActionOperation(weapon)
        )
        return character
    }
}
